dataland-e2etests/src/test/kotlin/org/dataland/e2etests/utils/ApiAccessor.kt

```diff
@@ -122,9 +127,38 @@ class ApiAccessor {
                 )
             }
         }
+        if (ensureQaPassed) ensureQaCompletedAndUpdateMetadata(listOfUploadInfo)
         return listOfUploadInfo
     }
 
+    /**
+     * Wait until QaStatus is accepted for all Upload Infos or throw error. The metadata of the provided uploadInfos
+     * are updated in the process.
+     *
+     * @param uploadInfos List of UploadInfo for which an update of the QAStatus should be checked and awaited
+     * @return Input list of UplaodInfo but with updated metadata
+     */
+    private fun ensureQaCompletedAndUpdateMetadata(uploadInfos: List<UploadInfo>) {
+        await().atMost(10, TimeUnit.SECONDS).until { checkIfQaPassedAndUpdateMetadata(uploadInfos) }
+    }
+
+    private fun checkIfQaPassedAndUpdateMetadata(uploadInfos: List<UploadInfo>): Boolean {
+        var allQaPassedIndicator = true
+        for (uploadInfo in uploadInfos) {
+            val metaData = uploadInfo.actualStoredDataMetaInfo
+                ?: throw NullPointerException(
+                    "To check QA Status, metadata is required but was null for $uploadInfo",
+                )
+            if (metaData.qaStatus != QAStatus.accepted) {
+                uploadInfo.actualStoredDataMetaInfo = metaDataControllerApi.getDataMetaInfo(metaData.dataId)
+                if (uploadInfo.actualStoredDataMetaInfo!!.qaStatus != QAStatus.accepted) {
+                    allQaPassedIndicator = false
+                }
+            }
+        }
+        return allQaPassedIndicator
+    }
```

Comment:
The Pattern "set a boolean to an initial value, and reverse it's value on a specific condition afterwards, and that's what you return" can be avoided (as it's not so nice) by using the streams-equivalent of Kotlin.

So this method can be simplified to:
```
    private fun checkIfQaPassedAndUpdateMetadata(uploadInfos: List<UploadInfo>): Boolean {
        return uploadInfos.all { uploadInfo ->
            val metaData = uploadInfo.actualStoredDataMetaInfo
                    ?: throw NullPointerException(
                            "To check QA Status, metadata is required but was null for $uploadInfo",
                    )
            if (metaData.qaStatus != QAStatus.accepted) {
                uploadInfo.actualStoredDataMetaInfo = metaDataControllerApi.getDataMetaInfo(metaData.dataId)
            }
            return uploadInfo.actualStoredDataMetaInfo!!.qaStatus == QAStatus.accepted
        }
    }
```

Probably it's even a little bit more ressource efficient as it stops processing once the first updated QA Status is != Accepted.

Reply:
since in Kotlin streams on Collections are eager instead of lazy my argument was wrong. I've adapted the proposed stream-based implementation.

Reply:
ok